### ISSUE TITLE
[BUG] cnmem.o linked twice.

### ISSUE THREAD

**Issue by seunghwak** [rapidsai/cugraph#251]
**Describe the bug**
cnmem is a pool memory allocator implemented in cnmem.cpp. cnmem.o is linked twice; through nvgraph and RMM.

See cugraph/cpp/nvgraph/cpp/CMakeLists.txt

```
108 if(NVGRAPH_LIGHT MATCHES True)
109     add_library(nvgraph_rapids SHARED
110                 thirdparty/cnmem/src/cnmem.cpp
111                 src/arnoldi.cu
112                 src/bfs.cu
...
142 else(NVGRAPH_LIGHT MATCHES True)
143         add_library(nvgraph_rapids SHARED
144                 thirdparty/cnmem/src/cnmem.cpp
145                 src/arnoldi.cu
146                 src/bfs.cu
...
```

RMM also internally uses cnmem.

libcugraph.so is dependent on libnvgraph_rapids.so and librmm.so and both libnvgraph_rapids.so and librmm.so include cnmem.o.

If we call RMM_ALLOC (macro) from cugraph, this calls rmm::alloc (defined in a header file), and rmm::alloc calls cnmemMalloc in libnvgraph_rapids.so.

If we call rmmAlloc (defined in rmm.cpp), this calls rmm::alloc and this rmm::alloc calls cnmemMalloc in librmm.so.

If we initialize cnmem through rmm, rmmAlloc calls succeed, but RMM_ALLOC calls fail because cnmem in libnvgraph_rapids.so is not initialized; and if both are properly initialized, there can be two memory pools and this is problematic.

We need to update nvgraph to use RMM (no direct use of cnmem).

**Comment by afender** [comment]
RMM was enabled in cugraph before 0.6 by @aschaffer (but not in nvgraph).

Nvgraph has been linking against cnmem long before rmm and cugraph were created. It is surprising that this conflict shows up just now.  

> if both are properly initialized, there can be two memory pools and this is problematic
What is error actually? Nvgraph should use its legacy memory pool and cugraph should use RMM.

> We need to update nvgraph to use RMM (no direct use of cnmem).
This is a lot of work and our efforts should be focused on cuGrpah rather than nvGraph. It is best to find out a workaround if possible imo.

**Comment by seunghwak** [comment]
RMM's pool allocator is currently not enabled by default. If pool allocator is not enabled, RMM calls are just wrappers to cudaMalloc/cudaFree, so no effect.

If we call RMM_ALLOC from cugraph, it eventually invokes cnmemAlloc in libnvgraph.so, not cnmemAlloc in librmm.so; and RMM_ALLOC does not work as expected.

If we want to fix this in a really dirty way, we may just rename functions in cnmem.cpp in nvgraph... but this does not sound like a proper fix if nvgraph is going to stay in cugraph for an extended period of time.

**Comment by seunghwak** [comment]
And if there are two memory pools, memory request on one pool can fail even if there is a plenty of available memory in another pool. cugraph's behavior can become really erratic once we mix regular cugraph calls and nvgraph calls.

**Comment by afender** [comment]
I see. There are two things:
1. Fix the conflict so that we can use RMM in cuGraph.
2. Upgrade nvgraph to support RMM so that there is a single memory pool.

For (2) I a bit reluctant to invest time in that because that would actually have limited benefits (nvgraph accepts RMM memory for I/O, the memory pool is just for internal allocations) and impact few features (for instance Pagerank, Jaccard, Louvain won't benefit from that).

**Comment by seunghwak** [comment]
If the consequence is less than ideal performance for rare use cases, I agree that we may not put too much effort on improving the performance. But if the consequence is erroneous program behavior, we should either fix it or drop the feature; we really cannot ask users to use a feature with a known bug (even if that happens only infrequently).

My current understanding is that this is a bug. If users make multiple nvgraph calls using big memory (and if nvgraph's cnmem takes all the GPU memory), any following cugraph calls relying on RMM's cnmem will fail. This is not just a performance hit but a bug and cannot be tolerated...

Let me take a closer look on this and what's the best way to fix this.

**Comment by seunghwak** [comment]
So we may compromise performance for rare use cases, but we cannot compromise correctness. And I think all the rapids projects should use a single pool allocator or no rapids project should use a pool locator for GPU memory allocation to enforce correctness (unless pool allocators can move memory blocks from one pool to another based on demand, this is unlikely to be a case anytime soon).

So we have two choices. 1) Replace the cnmem pool allocator in nvgraph with cudaMalloc/cudaFree, and in the cugraph's nvgraph wrapper, check whether RMM's pool allocation mode is enabled or not, and if enabled, throw an exception saying nvgraph doesn't work with RMM's pool allocation mode. 2) Update nvgraph to use RMM instead of cnmem.

If nvgraph is using just cudaMalloc/cudaFree, option 1 is much easier to implement, but in our case, implementing option 2 is not significantly more difficult than implementing option 1, so we may better go with option 2.

And I agree that implementing option 2 is not trivial, but I don't think this is a formidable amount of work.

Basically, nvgraph is using cnmem through

allocateDevice/attachDevicePtr
Vector (though this is not true resizable vector)
direct invocation of cnmemMalloc/cnmemFree

and in several places, nvgraph is directly invoking cudaMalloc/cudaFree.

So we need to replace these routines to use RMM. And we also need to check thrust calls if they involve memory allocations.

Please assign this work to me, and I will fix it.

@afender @BradReesWork 
One question: so I understand we should not update the nvgraph's public API as the main reason we are keeping nvgraph is for the legacy nvgraph users, but are we also trying to minimize changing nvgraph internals, or as long as the public API remains unchanged, do we have freedom to clean up the code internals?

**Comment by afender** [comment]
I think option 1) is less than a day of work but I'd agree that 2) is the ideal solution. Do you have a time estimate for option 2?

> And we also need to check thrust calls if they involve memory allocations.

 None is using the pool allocator afaik. I think it will take some time to update that. 

Regarding your question, Brad and I had a discussion about that yesterday. We are thinking about having two separate and independent versions of nvgraph. cuGraph should build only with its integrated version of nvGraph that we can freely modify. We will also have a legacy version that we only update with bug fixes. While having multiple code bases is not always preferred, it might be better in the long run in our case.

**Comment by seunghwak** [comment]
@afender I am thinking about several days.

and

```
cuGraph should build only with its integrated version of nvGraph that we can freely modify. We will also have a legacy version that we only update with bug fixes. 
```

This makes sense to me. We may go with option 1 for external packages we are not maintaining, but for nvgraph, I also think that option 2 is a better way to fix this problem.

**Comment by afender** [comment]
Ok. Let's add RMM in nvgraph in 0.8.

**Comment by afender** [comment]
Also, let's add a few tests in cugraph that are "really" running with RMM.
